### ISSUE TITLE
Appending audio to already recorded file

### ISSUE THREAD

**Issue by swaroopbutala** [syedhali/EZAudio#118]
Hello,

I am interested in appending data to already recorded file. I already have a file named "testfile.m4a"

When I try to use this recorder to read from microphone input, it overwrites the file. I would like to append to the file. Is there a way to do it?

self.recorder = EZRecorder(destinationURL: getFilePathUrl("testfile.m4a"), sourceFormat: self.microphone.audioStreamBasicDescription(), destinationFileType: EZRecorderFileType.M4A)

Thanks!

**Comment by swaroopbutala** [comment]
Thanks @lususvir! I was actually doing similar appending but was doing it using aiff format. m4a is much better with space! :)

**Comment by syrakozz** [comment]
+1  it will be great if it can merged with this project

**Comment by lususvir** [comment]
I've noticed that the appending process takes a long time. If you were making a recording app, your users would notice that it takes your app 5 to 10 seconds to append two files together, whereas Voice Memos does it instantaneously. 

Has anyone figured out how to simply record to the same file multiple times and just append audio at the end?

**Comment by lususvir** [comment]
Finally found a better way to do this. You should initialize your `EZRecorder` in `viewDidLoad:` when you press your record/pause toggle, it should really only set your flag `isRecording` to `YES` or `NO`. Then just surround everything in your EZMicrophone delegate with an if statement to see if you are recording:

I downloaded the EZRecorder example project and changed the following:

    -(void)toggleMicrophone:(id)sender {
       //I erased everything in here, I control the microphone with my record/pause button.

    }


    // In this method I never close the audio file, I only change the isRecording flag from YES to NO.
    // I also start fetching audio the first time record is pressed, and never stop until the user exits the recorder.
    -(void)toggleRecording:(id)sender {
  
    self.playingTextField.text = @"Not Playing";
    if( self.audioPlayer )
    {
        if( self.audioPlayer.playing )
        {
            [self.audioPlayer stop];
        }
        self.audioPlayer = nil;
    }
    self.playButton.hidden = NO;
    
    if( [sender isOn] )
    {
    //start the recorder
        self.isRecording = YES;
        [self.microphone startFetchingAudio];
        
        }
        else
        {
    //don't stop the recorder
            self.isRecording = NO;
        
        }
        self.isRecording = (BOOL)[sender isOn];
        self.recordingTextField.text = self.isRecording ? @"Recording" : @"Not Recording";
    }


Here is where you pause/unpause the actual recording and waveform drawing:

    -(void)microphone:(EZMicrophone *)microphone
     hasAudioReceived:(float **)buffer
     withBufferSize:(UInt32)bufferSize
    withNumberOfChannels:(UInt32)numberOfChannels {
    dispatch_async(dispatch_get_main_queue(),^{
      //add this flag to pause/unpause
      if(self.isRecording) {
          [self.audioPlot updateBuffer:buffer[0] withBufferSize:bufferSize];
      }
    });
    }

    -(void)microphone:(EZMicrophone *)microphone
    hasBufferList:(AudioBufferList *)bufferList
   withBufferSize:(UInt32)bufferSize
withNumberOfChannels:(UInt32)numberOfChannels {
  
     //I believe this flag was already there
      if( self.isRecording ){
      [self.recorder appendDataFromBufferList:bufferList
                             withBufferSize:bufferSize];
      }
  
    }

You can adapt this to your own UI. I was able to record, pause, record, pause, record, and pause, and all three recordings were appended together instantaneously without loading or delay. It was as good as the stock Voice Memos app. Maybe this was obvious to some, but no one answered back on here so I just now figured it out.

**Comment by KevinJacob** [comment]
@lususvir with your method it is no longer possible to play your recording till you close the file correct? If so do you know of anyway to play the recording between appends?

**Comment by lususvir** [comment]
I will have to check again, but I think I was able to play the audio file using AVAudioPlayer. I would think it wouldn't be possible, but I definitely played the file back to make sure it was appending and it played. So let me double check what I was doing.

**Comment by KevinJacob** [comment]
Ya I don't think it would. Since by not closing the recording you have no file to read in. I tired to use data buffers and load the NSData into AVAudioPlayer but that didn't work.

**Comment by lususvir** [comment]
Yeah, you're right. I could record multiple times and play it, but the audio player closes the file, so if you try to record after playing it, nothing happens.

**Comment by subinkk08** [comment]
i need to merge two files . i converted to swift code . 2nd audio is  not merging and exportSession.status is failed
can any one knows please help me 
 func combineTwoFile(session:Int) { 

if session > 0 { 
    let composition: AVMutableComposition = AVMutableComposition()

    var appendedAudioTrack: AVMutableCompositionTrack = composition.addMutableTrackWithMediaType(AVMediaTypeAudio, preferredTrackID: kCMPersistentTrackID_Invalid)

    let path1 = NSSearchPathForDirectoriesInDomains(
        .DocumentDirectory, .UserDomainMask, true
        ).first


   let orginalFile = NSURL.fileURLWithPath(path1! + "/EZAudioTest_0.m4a")
   let currentFile = NSURL.fileURLWithPath(path1! + "/EZAudioTest_"+session.description+".m4a")


    let originalAsset: AVURLAsset = AVURLAsset(URL: orginalFile, options: nil)
    let newAsset: AVURLAsset = AVURLAsset(URL: currentFile, options: nil)

    var tracks1 =  originalAsset.tracksWithMediaType(AVMediaTypeAudio)

    var tracks2 =  newAsset.tracksWithMediaType(AVMediaTypeAudio)

    var assetTrack1:AVAssetTrack = tracks1[0] as! AVAssetTrack
    var assetTrack2:AVAssetTrack = tracks2[0] as! AVAssetTrack

    var duration1: CMTime = assetTrack1.timeRange.duration
    var duration2: CMTime = assetTrack2.timeRange.duration

    var timeRange1 = CMTimeRangeMake(kCMTimeZero, duration1)
    var timeRange2 = CMTimeRangeMake(duration1, duration2)


    try? appendedAudioTrack.insertTimeRange(timeRange1, ofTrack: assetTrack1, atTime: kCMTimeZero)


    try? appendedAudioTrack.insertTimeRange(timeRange2, ofTrack: assetTrack2, atTime: duration1)

         let exportSession: AVAssetExportSession = AVAssetExportSession(asset: composition, presetName: AVAssetExportPresetAppleM4A)!
    if exportSession == false {
        // do something
        return
    }


    // make sure to fill this value in
    exportSession.outputURL = NSURL.fileURLWithPath(path1! + "/EZAudioTest_0.m4a")

    exportSession.outputFileType = AVFileTypeAppleM4A;
    exportSession.exportAsynchronouslyWithCompletionHandler({() -> Void in
        // exported successfully?
        switch exportSession.status {
        case .Failed: break

        case .Completed: break
            // you should now have the appended audio file

        case .Waiting: break

        default: break
        }
      //  var error: NSErrorPointer? = nil
    })
}
}
